pkg/backup/backup.go

```diff
@@ -167,101 +174,128 @@ func (b *Backup) Run() {
 		if ackchan != nil {
 			ack := backupNowAck{err: err}
 			if err == nil {
-				ack.status = b.getLatestBackupStatus()
+				ack.status = bc.backupManager.getLatestBackupStatus()
 			}
 			ackchan <- ack
 		}
 	}
 }
 
-func (b *Backup) saveSnap(lastSnapRev int64) (int64, error) {
-	podList, err := b.kclient.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+// saveSnap saves the latest snapshot if latest revision is greater than the given lastSnapRev.
+func (b *BackupManager) saveSnap(lastSnapRev int64) (int64, error) {
+	start := time.Now()
+
+	n, rev, etcdVersion, err := SaveSnap(b.kclient, b.etcdTLSConfig, lastSnapRev, b.be, b.namespace, b.clusterName)
 	if err != nil {
-		return lastSnapRev, err
+		return lastSnapRev, fmt.Errorf("failed to SaveSnap: (%v)", err)
 	}
 
-	var pods []*v1.Pod
-	for i := range podList.Items {
-		pod := &podList.Items[i]
-		if pod.Status.Phase == v1.PodRunning {
-			pods = append(pods, pod)
-		}
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
 	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	}
+
+	return rev, nil
+}
 
+// SaveSnap saves snapshot to backend from the etcd member whose rev is greater than lastSnapRev.
+func SaveSnap(kclient kubernetes.Interface, tls *tls.Config, lastSnapRev int64, be backend.Backend, namespace, clusterName string) (int64, int64, string, error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return 0, 0, "", fmt.Errorf("failed to retrieve running etcd pods: (%v)", err)
+	}
 	if len(pods) == 0 {
-		msg := "no running etcd pods found"
-		logrus.Warning(msg)
-		return lastSnapRev, fmt.Errorf(msg)
+		return 0, 0, "", fmt.Errorf("no running etcd pods found")
 	}
-	member, rev := getMemberWithMaxRev(pods, b.etcdTLSConfig, b.selfHosted)
-	if member == nil {
-		logrus.Warning("no reachable member")
-		return lastSnapRev, fmt.Errorf("no reachable member")
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return 0, 0, "", fmt.Errorf("no reachable member")
 	}
 
 	if rev <= lastSnapRev {
 		logrus.Info("skipped creating new backup: no change since last time")
-		return lastSnapRev, nil
-	}
-
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+		return 0, 0, "", fmt.Errorf("skipped creating new backup: no change since last time")
 	}
-	return rev, nil
-}
-
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
```

Comment:
why this func is removed? only the receiver needs to be changed to backupManager to achieve our goal, no?

Reply:
/cc @fanminshi  this needs to be resolved

Reply:
`only the receiver needs to be changed to backupManager to achieve our goal, no?`
possibly. let me think about how to use `backupManager` in the operator controller code.